espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/Source.java

```diff
@@ -0,0 +1,100 @@
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerConfigurationException;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
+import io.appium.espressoserver.lib.model.AppiumParams;
+import io.appium.espressoserver.lib.model.ViewAttributesEnum;
+import io.appium.espressoserver.lib.model.ViewElement;
+import io.appium.espressoserver.lib.viewaction.RootViewFinder;
+
+public class Source implements RequestHandler<AppiumParams, String> {
+
+    @Override
+    @Nullable
+    public String handle(AppiumParams params) throws AppiumException {
+
+        try {
+            // Build a document
+            Document doc;
+            DocumentBuilderFactory docFactory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder docBuilder = docFactory.newDocumentBuilder();
+            doc = docBuilder.newDocument();
+
+            // Populate the XML
+            View rootView = (new RootViewFinder()).getRootView();
+            buildXML(doc, null, rootView);
+
+            // Write the XML to string
+            TransformerFactory transformerFactory = TransformerFactory.newInstance();
+            Transformer transformer = transformerFactory.newTransformer();
+            DOMSource source = new DOMSource(doc);
+            StringWriter stringWriter = new StringWriter();
+            StreamResult result = new StreamResult(stringWriter);
+            transformer.transform(source, result);
+            return stringWriter.toString();
+
+        } catch (ParserConfigurationException pe) {
+            throw new AppiumException(String.format("Could not parse XML from source: %s", pe.getMessage()));
+        } catch (TransformerConfigurationException te) {
+            throw new AppiumException(String.format("Could not parse XML from source: %s", te.getMessage()));
+        } catch (TransformerException te) {
+            throw new AppiumException(String.format("Could not parse XML from source: %s", te.getMessage()));
+        }
+    }
+
+    /**
+     * Recursively visit all of the views and map them to XML elements
+     * @param doc XML Document
+     * @param parentElement Element that this new element will be appended to
+     * @param view Android View that will map to an Element
+     */
+    public void buildXML(Document doc, Element parentElement, View view) {
+        // TODO: Extract this logic out to it's own class so it can be re-used with XPath locators
+        Element element = doc.createElement(view.getClass().getName());
+
+        // Set attributes
+        ViewElement viewElement = new ViewElement(view);
+        element.setAttribute(ViewAttributesEnum.CONTENT_DESC.getName(), viewElement.getContentDescription());
```

Comment:
I find this is not very safe, since XML standard explicitly defines the set of acceptable characters. It would be better to have helper methods, which verifies the input and replaces unacceptable characters with empty chars.
Hint: you can copy such method from uiautomator2-server source (the name is **stripInvalidXMLChars**).